src/main/java/org/apache/accumulo/testing/continuous/BulkKey.java

```diff
@@ -0,0 +1,236 @@
+package org.apache.accumulo.testing.continuous;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+
+import org.apache.accumulo.core.data.ByteSequence;
+import org.apache.accumulo.core.data.Key;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.io.WritableComparable;
+import org.apache.hadoop.io.WritableComparator;
+import org.apache.hadoop.io.WritableUtils;
+
+/**
+ * Supports immediate sorting via eager deser of the key object. This has the benefit of reducing
+ * the amount of deserialization that may occur when sorting keys in memory
+ */
+public class BulkKey implements WritableComparable<BulkKey> {
+
+  protected Key key = new Key();
+  protected int hashCode = 31;
+
+  static final byte[] EMPTY = {};
+
+  Text row = new Text();
+  Text cf = new Text();
+  Text cq = new Text();
+  Text cv = new Text();
+
+  public BulkKey() {}
+
+  public BulkKey(Key key) {
+    this.key = key;
+    hashCode = key.hashCode();
+  }
+
+  public BulkKey(byte[] row, byte[] cf, byte[] cq, byte[] cv, long ts, boolean deleted) {
+    // don't copy the arrays
+    this.key = new Key(row, cf, cq, cv, ts, deleted, false);
+    hashCode = key.hashCode();
+  }
+
+  public Key getKey() {
+    return key;
+  }
+
+  public ByteSequence getRowData() {
+    return key.getRowData();
+  }
+
+  @Override
+  public void readFields(DataInput in) throws IOException {
+
+    final int rowsize = WritableUtils.readVInt(in);
+    final byte[] row = readBytes(in, rowsize);
+
+    final int cfsize = WritableUtils.readVInt(in);
+    final byte[] cf = readBytes(in, cfsize);
+
+    final int cqsize = WritableUtils.readVInt(in);
+    final byte[] cq = readBytes(in, cqsize);
+
+    final int cvsize = WritableUtils.readVInt(in);
+    final byte[] cv = readBytes(in, cvsize);
+
+    final long ts = WritableUtils.readVLong(in);
+    boolean isDeleted = in.readBoolean();
+
+    key = new Key(row, cf, cq, cv, ts, isDeleted, false);
+
+    hashCode = key.hashCode();
+  }
+
+  private static byte[] readBytes(DataInput in, int size) throws IOException {
+    if (size == 0)
+      return EMPTY;
+    final byte[] data = new byte[size];
+    in.readFully(data, 0, data.length);
+    return data;
+  }
+
+  @Override
+  public void write(DataOutput out) throws IOException {
+
+    key.getRow(row);
+    key.getColumnFamily(cf);
+    key.getColumnQualifier(cq);
+    key.getColumnVisibility(cv);
+
+    WritableUtils.writeVInt(out, row.getLength());
```

Comment:
For the bulk ingest data, the row, family, and qualifier are all really small.  The length of each of these should fit in a single byte.  Could just always store the len in a single byte here instead of using writeVInt.  Seems like this  could simplify the code in KeyShortCircuitComparator.

Could have a little util method like 

```java
  void writeLen(OutputStream out, int len){
      Preconditions.checkArgument(len<128);
      out.write(len);
  }
```

Reply:
Could take this even further and assume the row, family, and qual are always the same len (with sanity checks) and not bother serializing the len.  This could make the compare in KeyShortCircuitComparator really simple and super fast (would not even need to worry about fields).

Reply:
Good ideas. Thanks!

Reply:
@phrocker  curious if you ever got a chance to give this a try.

Reply:
@keith-turner I'm iteratively burning down my PRs. I will post updates to this on Friday with some testing. I have a sizable cluster where I can run this and post timing then. thanks!

Reply:
I was able to try it @keith-turner . Made a positive impact! Thanks! I'm actually adding some unit tests in the Key class, so I'll shoot a rebased and updated  PR Monday morning since it's gettin a bit late. This one was fun to test. 

FYI I took your suggestions of reducing the complexity of the key comparator and removing the timestamp/delete. I added some unit tests that function as regression tests in case any assumptions change in the testing framework. BulkKey was kind of general, but "TestKey" is very specific and focused on the specificities of size/type of keys.

Reply:
Awesome.  Yeah a more specific name for the class would be good.  

> I added some unit tests that function as regression tests in case any assumptions change in the testing framework.

Nice, much better than runtime sanity checks.

Reply:
One thing I did not change was writing V int that's because it writes smaller sizes. One issue I ran into was spilling. If I use writeInt it writes 4 bytes, whereas writeVInt writes 1 byte for our sizes. While there are more branches it writes significantly less data to the context, resulting in overall greater throughput.

Reply:
I think that saved 12 bytes per key, which was enough to keep me from avoiding a second spill and drastically slowing my jobs down.